File: lc_soln/0543_diameter_of_binary_tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        result = [0]
        self.dfs(root, result)
        return result[0]

    def dfs(self, root: Optional[TreeNode], result: list[int]) -> int:
        if root is None:
            return 0
        left = self.dfs(root.left, result)
        right = self.dfs(root.right, result)
        result[0] = max(result[0], left + right)
        return 1 + max(left, right)


def main(): ...


if __name__ == "__main__":
    main()

File: lc_soln/0543_diameter_of_binary_tree/solution.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution
{
    public int DiameterOfBinaryTree(TreeNode root)
    {
        int result = 0;
        Dfs(root, ref result);
        return result;
    }

    int Dfs(TreeNode root, ref int result)
    {
        if (root is null)
        {
            return 0;
        }
        int left = Dfs(root.left, ref result);
        int right = Dfs(root.right, ref result);
        result = int.Max(result, left + right);
        return 1 + int.Max(left, right);
    }
}
